src/epik_eval/dataloader.py

```diff
@@ -0,0 +1,205 @@
+        segmented_document_directory (str): Path to the directory containing the segmented stories. Default: 'data/dataset/segmented_documents/'
+        eval_questions_csv (str): Path to the evaluation (val & test) questions. Default: 'data/dataset/eval_questions.csv'
+    """
+
+    def __init__(
+            self,
+            tokenizer,
+            batch_size,
+            model_type='causal',
+            segmented_docs=False,
+            unsegmented_document_directory='data/dataset/unsegmented_documents/',
+            segmented_document_directory='data/dataset/segmented_documents/',
+            eval_questions_csv='data/dataset/eval_questions.csv'):
+        super().__init__()
+        assert model_type in ['causal', 't5'], f"Expected model_type to be either 'causal' or 't5', but got: {model_type}"
+
+        qa_df = pd.read_csv(eval_questions_csv, dtype=str, delimiter='|')
+        questions = qa_df['question'].to_list()
+        answers = qa_df['answer'].to_list()
+
+        # Model is also evaluated for its ability to memorize training samples.
+        train_doc_ids = []
+        train_doc_tasks = []
+        train_doc_banners = []
+        train_doc_stories = []
+        document_directory = segmented_document_directory if segmented_docs else unsegmented_document_directory
+        for file in sorted(os.listdir(document_directory)):
+            if file[-4:] == '.txt':
+                with open(os.path.join(document_directory, file), 'r') as f:
+                    doc = f.read().split('\n')
+                    train_doc_ids.append(file[:4])
+                    train_doc_tasks.append(qa_df.loc[qa_df['story id'] == file[:4]]['task'].item())
+                    train_doc_banners.append(doc[0])
+                    if model_type == 't5': # T5 tokenizer converts \n into spaces, so must replace \n with space
+                        train_doc_stories.append(' '.join(doc[1:]))
+                    else:
+                        train_doc_stories.append('\n'.join(doc[1:]))
+
+        self.prompts = questions + train_doc_banners
+        self.targets = answers + train_doc_stories
+
+        # Prepare batches once since always the same
+        self.batches = []
+        for i in range(0, len(self.prompts), batch_size):
+            batch = tokenizer(self.prompts[i: i + batch_size], padding=True, return_tensors='pt')
+            prompt_ids = range(i, min(i + batch_size, len(self.prompts)))
+            self.batches.append((prompt_ids, batch.input_ids, batch.attention_mask))
+
+        train_df = pd.DataFrame({
+            'set': ['train'] * len(train_doc_ids),
+            'story id': train_doc_ids,
+            'task': train_doc_tasks,
+            'question': train_doc_banners,
+            'answer': train_doc_stories
+        })
+        self.eval_df = pd.concat([qa_df, train_df], ignore_index=True)
+
+        self.current_batch = 0
+        self.finished = False
+
+    def __iter__(self):
+        if self.finished:
```

Comment:
An `Iterable` does not finish, it is stateless. An `iterator` finish. I think there is a mix of responsibility here, which is properly what is causing you issues.

I guess, the iterator that this __iter__ returns, should either yield all observations from every batch or it should yield all batches.

Reply:
See my comment below explaining the design.